homeassistant/components/saj/sensor.py

```diff
@@ -0,0 +1,161 @@
+    """Set up SAJ sensors."""
+
+    # Init all sensors
+    sensor_def = pysaj.Sensors()
+
+    # Use all sensors by default
+    hass_sensors = []
+    used_sensors = []
+
+    for sensor in sensor_def:
+        hass_sensors.append(SAJsensor(sensor))
+        used_sensors.append(sensor)
+
+    saj = pysaj.SAJ(config[CONF_HOST])
+
+    async_add_entities(hass_sensors)
+
+    backoff = 0
+    backoff_step = 0
+
+    async def async_saj(event):
+        """Update all the SAJ sensors."""
+        nonlocal backoff, backoff_step
+        tasks = []
+        # If reading sensors has failed multiple times
+        if backoff > 1:
+            backoff -= 1
+
+            # SAJ inverters are powered by DC via solar panels and thus are
+            # offline after the sun has set. If a sensor resets on a daily
+            # basis like "today_yield", this reset won't happen automatically.
+            # Code below checks if today > day when sensor was last updated
+            # and if so: set state to STATE_UNKNOWN.
+            # Sensors with live values like "temperature" or "current_power"
+            # will also be reset to STATE_UNKNOWN.
+            for sensor in hass_sensors:
+                if (sensor.per_day_basis and date.today() > sensor.date_updated) or (
+                    not sensor.per_day_basis and not sensor.per_total_basis
+                ):
+                    task = sensor.async_update_values(unkown_state=True)
+                    if task:
+                        tasks.append(task)
+            if tasks:
+                await asyncio.wait(tasks)
+
+            return
+
+        values = await saj.read(used_sensors)
+        if not values:
+            try:
+                backoff = [1, 1, 1, 6, 30][backoff_step]
+                backoff_step += 1
+            except IndexError:
+                backoff = 60
+            return
+        backoff_step = 0
+
+        for sensor in hass_sensors:
+            task = sensor.async_update_values()
+            if task:
+                tasks.append(task)
+        if tasks:
+            await asyncio.wait(tasks)
+
+    interval = config.get(CONF_SCAN_INTERVAL) or timedelta(seconds=5)
+    async_track_time_interval(hass, async_saj, interval)
+
+
+class SAJsensor(Entity):
+    """Representation of a SAJ sensor."""
+
+    def __init__(self, pysaj_sensor):
+        """Initialize the sensor."""
+        self._sensor = pysaj_sensor
+        self._state = self._sensor.value
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"saj_{self._sensor.name}"
+
+    @property
+    def icon(self):
+        """Return the icon for the sensor."""
+        return SENSOR_ICON
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._sensor.unit
```

Comment:
What is the unit we're measuring? Can we set device class power?

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

Reply:
The measure unit depends on which sensor, the full list is available over here: https://deploy-preview-10455--home-assistant-docs.netlify.com/components/saj/
Is it ok that I check the `self._sensor.unit` value if it corresponds with `W` or `kW` to set the device class to `power`, and if it's `°C` set device class to `temperature` ?

Can `kWh` also belong to device class `power` ?

Reply:
Yes, it's ok to check like that. But we should use the home assistant constants when we return the unit of measurement here, for the units where we have those, eg `POWER_WATT`.

https://github.com/home-assistant/home-assistant/blob/1c72a246a048d0f2f3d3dca0b77be559dae01a78/homeassistant/const.py#L325

No, `kWh` isn't part of the specified units allowed for that device class according to our dev docs.

Reply:
@MartinHjelmare will the hass frontend automatically convert W to kW or kW to kWh? In many of these sensors the less significant digits just clutters the UI

Reply:
(Also might be difficult to use hass constants if these devices are defined in a library..)

Reply:
I don't know about the frontend here. The backend doesn't care about the frontend.

Library constants should preferably be exposed so they can be imported and mapped to home assistant constants.

Reply:
I agree with @kellerza that if the unit of measurement is defined in a library that it's difficult to use HASS constants to display them.
What I did do is added device class when necessary and check them against constants.
```
    @property
    def device_class(self):
        """Return the device class the sensor belongs to."""
        if self._sensor.unit == POWER_WATT:
            return DEVICE_CLASS_POWER
        if self._sensor.unit == TEMP_CELSIUS or self._sensor.unit == TEMP_FAHRENHEIT:
            return DEVICE_CLASS_TEMPERATURE
```

Reply:
> Library constants should preferably be exposed so they can be imported and mapped to home assistant constants.

Apparently we've replied at the same time so I've missed your answer...
So I should create a mapping between the string values the library returns and the HASS constants that do exist?

Reply:
Yes, for more than a couple of cases, the most effective way to couple the APIs is a dict that we can look up.

Reply:
I've created a dict to map the lib unit of measurements with the hass constants.
The only one I did not find was one for "h" (hours) so I created a constant in the SAJ component for this.